### PR TITLE
WIP: Support black lenticular bracketed or square brackted group, title and episode number in parser

### PR DESCRIPTION
**Pull request by whtsky**
#### Database Migration
NO

#### Description
Added a new regex for parsing black lenticular bracketed or square brackted group, title and episode number.
All unit test in `AbsoluteEpisodeNumberParserFixture` has passed.

#### Todos
- [x] Tests
- [ ] Documentation

#### Issues Fixed or Closed by this PR

* #2898

### CODE REVIEW

**Comment by thezoggy** [comment]
i assume these are only found on a torrent site? does their api serve up multibyte characters?

**Comment by whtsky** [comment]
> does their api serve up multibyte characters?

They don't have any api; CJK charaters are used in RSS feed.
https://share.dmhy.org/topics/rss/rss.xml

**Comment by Taloth** [comment]
Probably easier to replace the special characters with [ ] as cleanup step in the parser (we have several of those), instead of adding another regex. Provided that one of the existing regexes will parse the result properly.

File: src/NzbDrone.Core.Test/ParserTests/QualityParserFixture.cs [hunk 1]
```diff
@@ -119,6 +119,7 @@ public void should_parse_webdl480p_quality(string title, bool proper)
         [TestCase("Survivorman.The.Lost.Pilots.Summer.HR.WS.PDTV.x264-DHD", false)]
         [TestCase("Victoria S01E07 - Motor zmen (CZ)[TvRip][HEVC][720p]", false)]
         [TestCase("flashpoint.S05E06.720p.HDTV.x264-FHD", false)]
+        [TestCase("【极影字幕社】 ★1月新番【巴哈姆特之怒：马纳利亚朋友】【Shingeki no Bahamut Manaria Friends】【01】GB MP4_720", false)]
```
Comment (Taloth) [inline src/NzbDrone.Core.Test/ParserTests/QualityParserFixture.cs]: Do you have other examples of this? Coz to me it just looks like a bad release name. We can't accept just '720' which is season 7 episode 20 in some cases, or 720 abs nr for certain korean shows.

File: src/NzbDrone.Core.Test/ParserTests/AbsoluteEpisodeNumberParserFixture.cs [hunk 2]
```diff
@@ -94,6 +94,15 @@ public class AbsoluteEpisodeNumberParserFixture : CoreTest
         [TestCase("Tokyo Ghoul: RE S2 - Episode 4 VOSTFR (1080p)", "Tokyo Ghoul RE S2", 4, 0, 0)]
         [TestCase("To Aru Majutsu no Index III - Episode 5 VOSTFR (1080p)", "To Aru Majutsu no Index III", 5, 0, 0)]
         [TestCase("[Prout] Steins;Gate 0 - Episode 5 VOSTFR (BDRip 1920x1080 x264 FLAC)", "Steins;Gate 0", 5, 0, 0)]
+        [TestCase("【喵萌奶茶屋】★10月新番★[哥布林杀手/Goblin Slayer][12END][720p][繁体][招募翻译校对]", "Goblin Slayer", 12, 0, 0)]
+        [TestCase("[爱恋字幕社][盾之勇者成名录][Tate no Yuusha no Nariagari][02][720P][GB][MP4]", "Tate no Yuusha no Nariagari", 2, 0, 0)]
```
Comment (Taloth) [inline src/NzbDrone.Core.Test/ParserTests/AbsoluteEpisodeNumberParserFixture.cs]: Which of these can Sonarr ACTUALLY process?

File: src/NzbDrone.Core/Parser/Parser.cs [hunk 3]
```diff
@@ -239,7 +243,7 @@ public static class Parser
                           RegexOptions.IgnoreCase | RegexOptions.Compiled),
 
                 //Anime - Title Absolute Episode Number
-                new Regex(@"^(?:\[(?<subgroup>.+?)\][-_. ]?)?(?<title>.+?)(?:[-_. ]+(?<absoluteepisode>(?<!\d+)\d{2,3}(\.\d{1,2})?(?!\d+)))+(?:_|-|\s|\.)*?(?<hash>\[.{8}\])?(?:$|\.)?",
+                new Regex(@"^(?:\[(?<subgroup>.+?)\][-_. ]?)?.*?(?<title>[a-z][\x20-\x5A\x5E-\x7F]+?)(?:[-_. ]+(?<absoluteepisode>(?<!\d+)\d{2,3}(\.\d{1,2})?(?!\d+)))+(?:_|-|\s|\.)*?(?<hash>\[.{8}\])?(?:$|\.)?",
```
Comment (Taloth) [inline src/NzbDrone.Core/Parser/Parser.cs]: This doesn't seem correct. `[a-z][\x20-\x5A\x5E-\x7F]+?` should only match a starting a-z.

**Comment by Taloth** [comment]
Like half the regexes in the parser are for anime, just not Chinese ones.

As mentioned, I'd like to know the actually usability for Chinese releases, since a few of the parser testcases cannot be processed automatically anyway due to various reasons.

**Comment by thezoggy** [comment]
seems like the [chinese] stuff should just stripped/ignored...  but then you have to deal with the other parts which just isnt supported naming.. so really its back to badly named releases that only exist on this one torrent site...

File: src/NzbDrone.Core.Test/ParserTests/QualityParserFixture.cs (hunk 1 above)
Comment (whtsky) [reply src/NzbDrone.Core.Test/ParserTests/QualityParserFixture.cs]: Looks like they are the only releasing group using this format.. I'll drop this commit

**Comment by whtsky** [comment]
> I'd like to know the actually usability for Chinese releases

It's not usable for Chinese anime releases. I found 0 working example right now.

> so really its back to badly named releases that only exist on this one torrent site...

It's not about torrent site, it's about the publisher.. Take a look at https://share.dmhy.org/topics/list/sort_id/2 , https://bangumi.moe/ and https://share.acgnx.se/sort-1-1.html . 

> seems like the [chinese] stuff should just stripped/ignored...

There's only a small (20% maybe) set of rls group put Chinese title in separated brackets.. The more common case is `[Chinese / English]`, `[Chinese_English]` and `[ChineseEnglish]` ( no separator )

File: src/NzbDrone.Core.Test/ParserTests/AbsoluteEpisodeNumberParserFixture.cs (hunk 2 above)
Comment (whtsky) [reply src/NzbDrone.Core.Test/ParserTests/AbsoluteEpisodeNumberParserFixture.cs]: Oh.. Just found that Sonarr will strip all non-ASCII charaters in release title

**Comment by whtsky** [comment]
So Sonarr strips all non-ASCII characters in title, which should make the parsing a lot easier than I initially thought. 

HELP WANTED: could you point out where did Sonarr did the strip? I want to replace `【】` to `[]` but couldn't find where to tweak

**Comment by whtsky** [comment]
I’m closing this PR since non-ASCII characters are striped before parseTitle ..

**Comment by Taloth** [comment]
sry mate, my work week is long and chaotic right now.

I suspect you don't want to pursue this feature any further, but I wanted to comment on your questions and remarks regardless:

The unicode appears to be stripped in RssParser, calling XmlCleaner.ReplaceUnicode, it would be feasible to perform common substitutions there, particularly the brackets.

> It's not usable for Chinese anime releases. I found 0 working example right now.

That's what I was afraid of, which means you'd need mapping on TheXEM to make certain titles work. Which begs the question on whether it's worth it.
If you believe there's a market for releases like this, then it might be possible to perform more substitutions on the rss level.

> There's only a small (20% maybe) set of rls group put Chinese title in separated brackets.. The more common case is [Chinese / English], [Chinese_English] and [ChineseEnglish] ( no separator )

Sounds to me like the parser should then have a regex to handle `[/?(English)]` titles.

**Pull request closed without merging**